### DISCUSSION TITLE
How to stop Vike from injecting `preload` tags client-side to the `<head>` of the document?

### DISCUSSION THREAD

**Discussion by ChristophP** [vikejs/vike#1819] [Discussion] [answered]
Hi,

## Short story

When using `injectFilter` to disable asset preloading, there are still preload link tags inserted by JavaScript once the page has loaded in the browser. Could this be a bug or is there some other way to remove them?

## Long story

We at BurdaForward have a special case where we want to provide a barebones page with just our layout (header & footer) to third party content providers. Since that template is meant to be hosted under a different domain we need the src/href attributes for the  assets and links in the HTML to point to be absolute and not relative (as to point back to `www.focus.de` instead of `thirdparty.focus.de`)

### The clean way

There are a few options to achieve something in this direction but that don't work well for us because they aren't runtime solutions like base assets options from https://vike.dev/base-url:
These are all build options. But our system is available under different domains (www.chip.de/www.focus.de/www.bunte.de). Using relative urls allowed us to use the same build for all brands. If we need to use a build option to turn things like `<img src="/path/to/image.jpg">` into `<img src="https://www.focus.de/path/to/image.jpg">` it would force use to use one build per brand :-/

### The dirty way

So to have a runtime solution we decided to do some post processing on the rendered HTML returned by Vike's `renderPage()` in the controller of our express app. Basically we parse the HTML and prefix all relative urls of certain tags with the base url of the current brand (performance is not a great concern for this case)
It works well BUT there still are some `preload` tags injected by Vike client side to the `<head>` of the document which contain relative urls (of course because Vike does not know about our rewriting mess).
We tried something like this to get rid of it

```ts
  const injectFilter = (assets: InjectFilterEntry[]): void => {
    assets.forEach((asset) => {
      asset.inject = false;
    });
  };
```
but the preload tags in the `<head>` continued to show up.

We also noticed that the preload tags showed up twice in the html.
1. once before the closing `</body>` which is server renderered
2. another time in the `<head>` which is client-side injected

Here's an example: https://www.focus.de/panorama/aus-aller-welt/katze-stiehlt-kleidung-von-nachbarn-zeigt-keine-anzeichen-damit-aufzuhoeren_39a216db-0863-437d-805b-e721f2dc383d.html
you can see one set of preload tags in the bottom (server rendered) and one set of preload tags injected at the top when looking in the dev tools.

## Conclusion

Would be great to know if removing those preload tags can be achieved.

**Comment by brillout** [comment]
Hi Dido,

> When using `injectFilter` to disable asset preloading, there are still preload link tags inserted by JavaScript once the page has loaded in the browser. Could this be a bug or is there some other way to remove them?

Yes, it seems to be a bug. Can you publish a reproduction?

> Here's an example: https://www.focus.de/panorama/aus-aller-welt/katze-stiehlt-kleidung-von-nachbarn-zeigt-keine-anzeichen-damit-aufzuhoeren_39a216db-0863-437d-805b-e721f2dc383d.html you can see one set of preload tags in the bottom (server rendered) and one set of preload tags injected at the top when looking in the dev tools.

That's expected: the preload tags for fonts are injected early in `<head>` whereas the preload tags for scripts are injected later at the end of `<body>`. There doesn't seem to be any preload tag duplication in that example:

```js
console.log(Array.from(document.querySelectorAll('[rel=preload]')).map(e => e.href).join('\n'))
console.log(Array.from(document.head.querySelectorAll('[rel=preload]')).map(e => e.href).join('\n'))
console.log(Array.from(document.querySelectorAll('[rel=modulepreload]')).map(e => e.href).join('\n'))
console.log(Array.from(document.body.querySelectorAll('[rel=modulepreload]')).map(e => e.href).join('\n'))
```

> ## Long story

I'm open to discuss solutions for your Base URL issue, although it seems like it would require a substantial new feature. As you rightfully pointed out, the Base URL is currently determined at build-time and there isn't a way to have dynamic/runtime Base URLs.

Since we're currently focusing on releasing the `v1.0.0` and on rehauling Vike's marketing, I'd suggest we try to solve the issue using the dirty approach first and see how it goes.

Keep me updated, and I'm looking forward to the reproduction.

**Comment by ChristophP** [comment]
Hi Rom,

thanks for your prompt reply.

## 1. Getting rid of preload tags

> Yes, it seems to be a bug. Can you publish a reproduction?

I created a repro for this behavior in this repo: https://github.com/ChristophP/vike-preload-repro .

It's pretty much a standard server-rendered Vike app.
The interesting part [is here](https://github.com/ChristophP/vike-preload-repro/blob/master/renderer/%2BonRenderHtml.tsx#L56-L62) where we filter out all assets to be injected.

Then  once we inspect the elements in the page after page load we can see that the `<body>` has no preload tags but the `<head>` does have 2. Running this short script in the browser shows that.

```js
console.log(Array.from(document.head.querySelectorAll('[rel="modulepreload"]')).map(e => e.href));
console.log(Array.from(document.body.querySelectorAll('[rel="modulepreload"]')).map(e => e.href));
```

Would you say this is a bug?

## 2. Duplication of tags

> That's expected: the preload tags for fonts are injected early in <head> whereas the preload tags for scripts are injected later at the end of <body>. There doesn't seem to be any preload tag duplication in that example:

You were right, the page I linked you didn't show that duplication of the preload tags. I had assumed it would happen on all pages when I gave you the public link, but it turns out it only happens when we rewrite urls which I only had locally.

If you want to see this behavior by
- turning on [this boolean condition here](https://github.com/ChristophP/vike-preload-repro/blob/master/server/index.ts#L74-L77), which will start the rewriting
- and commenting out the `injectFilter` section [here](https://github.com/ChristophP/vike-preload-repro/blob/master/renderer/%2BonRenderHtml.tsx#L56-L62)
- Then load a page
In this case we will get the tags once in the body AND once it the head (the body with absolute urls and the head with relative ones). It seems like Vike might have some logic like "if the preload tags are not there yet in the server renderered page, then inject them client side". You can run the script above again to see it.

I wouldn't call this behavior a Vike bug since what we're doing by rewriting urls is sort of weird 😅 but I since maybe this is related I wanted to bring it up.

## Other

> I'm open to discuss solutions for your Base URL issue, although it seems like it would require a substantial new feature. As you rightfully pointed out, the Base URL is currently determined at build-time and there isn't a way to have dynamic/runtime Base URLs. Since we're currently focusing on releasing the v1.0.0 and on rehauling Vike's marketing, I'd suggest we try to solve the issue using the dirty approach first and see how it goes.

Yeah, I think if we can make this approach work it would be a good-enough solution for now. I can see how allowing baseUrls at runtime would have many implications on a lot of places.
Looking forward to Vike v1.0.0, I see a lot of good developments with Vike that I like. Custom settings are a great thing to have 🙌

**Reply by CapCaWork** [reply]
Thanks @ChristophP ! 
@brillout is there anything we can do to support you here?

**Reply by brillout** [reply]
I didn't forget about this, but thanks for the ping! Let me finish what my current priorities and I'll then have a look at the reproduction. ETA today.

> @brillout is there anything we can do to support you here?

Yes, [company sponsorship](https://github.com/vikejs/vike/issues/1350) is the most effective and profound way to help. It makes a massive difference for us.

**Reply by brillout** [reply, marked as answer]
> ## 1. Getting rid of preload tags

> It's pretty much a standard server-rendered Vike app. The interesting part [is here](https://github.com/ChristophP/vike-preload-repro/blob/master/renderer/%2BonRenderHtml.tsx#L56-L62) where we filter out all assets to be injected.
> 
> Then once we inspect the elements in the page after page load we can see that the `<body>` has no preload tags but the `<head>` does have 2. Running this short script in the browser shows that.
> 
> ```js
> console.log(Array.from(document.head.querySelectorAll('[rel="modulepreload"]')).map(e => e.href));
> console.log(Array.from(document.body.querySelectorAll('[rel="modulepreload"]')).map(e => e.href));
> ```

[These are injected by Vite](https://github.com/vikejs/vike/commit/7a18e2acbed50c978a27cac02a0c084deaa905b6). You also need to set Vite's [`build.modulePreload`](https://vitejs.dev/config/build-options#build-modulepreload) setting to `false` if you want to remove all preload tags.

> ## 2. Duplication of tags

I think the duplication is caused by both Vike and Vite injecting preload tags. Correct me if I'm wrong, but I guess that if you disable Vite's injection then you'll get rid of the duplication.

> ## Other

> Yeah, I think if we can make this approach work it would be a good-enough solution for now.

:+1: Keep us updated.

Apologies for the late response, I was taking a little bit of a break and I was focused on finishing a couple of priorities.

**Comment by ChristophP** [comment]
@brillout thanks for looking into this.
Apologies for my late response as I just got back from vacation this week.

> [These are injected by Vite](https://github.com/vikejs/vike/commit/7a18e2acbed50c978a27cac02a0c084deaa905b6). You also need to set Vite's [build.modulePreload](https://vitejs.dev/config/build-options#build-modulepreload) setting to false if you want to remove all preload tags.

Good to understand that it is vite who does this. The bad news for us is that this is a build option which is not really a good solution for us.

In the end I think the problem is even trickier than this, because we found out that vite injects `import`ed styles client-side and there doesn't seem to be a way around it other than using `?inline` on all of the style imports. So seems like the sort of thing we're trying to do (rewriting urls server side) doesn't really play well with Vite's client-side injections.

**Reply by brillout** [reply]
No worries and let me know if there is anything Vike can do to help! Kind regards from Bogenhausen. (Btw. Vike's team member @magne4000 is a very experienced and up for consulting in case it's something you'd be up for.)

**Reply by magne4000** [reply]
Hi :wave: I'm indeed available for consulting if you need too.

> The bad news for us is that this is a build option which is not really a good solution for us.

You would need `modulePreload` parameter equivalent at runtime is that it?

> In the end I think the problem is even trickier than this, because we found out that vite injects imported styles client-side and there doesn't seem to be a way around it other than using ?inline on all of the style imports. So seems like the sort of thing we're trying to do (rewriting urls server side) doesn't really play well with Vite's client-side injections.

I'm wondering if a custom Vite plugin could do the trick here. Not sure though, it would need to be tested first.

**Comment by ChristophP** [comment]
@brillout @magne4000 Thanks for reaching out.

> You would need modulePreload parameter equivalent at runtime is that it?

That would have helped but understandably this is not something that sounds easy to do. Since I learned that some of the behavior that's causing problems is coming from Vite rather than Vike (such as Vite injecting styles via client-side JS) we kind of had to accept that doing a runtime rewite of relative to absolute urls is super tricky.

Using absolute urls was actually not the problem but the solution to a very specific problem of third party providers not wanting to host assets themselves but using out system instead. For now, it looks like we'll try to solve this problem a different way most likely by using a proxy in front of the 3rd party system that forwards assets calls to our system. With this solution absolute urls are not necessary anymore.

Anyway thank you for providing some insight into the nature of the problem. It helped understand the restrictions better.

**Reply by brillout** [reply]
:+1: I'm glad you got it resolved on your end. Let us know if you need anything from us.

**Reply by ChristophP** [reply]
Thank you. Much appreciated 🙏
